src/service/unauthorised.handler.service.ts

```diff
@@ -0,0 +1,41 @@
+import { Session } from "@companieshouse/node-session-handler";
+import { Request } from "express";
+
+import { logger } from "../utils/logger";
+import { createOAuthApiClient } from "./api.service";
+import { refreshToken } from "./refresh.token.service";
+
+/**
+ * Unauthorised response handler for the Update and Create OE calls.
+ * Retry the call after refreshing the token in the event of 401 unauthorised response.
+ *
+ * @param fnName Function Name (postOverseasEntity or putOverseasEntity)
+ * @param req Request object
+ * @param session Session object
+ * @param otherParams Parameters passed to the callback:
+ *        For the postOverseasEntity call we have (transactionID, appData, isSaveAndResumeFeatureActive)
+ *        For the putOverseasEntity call we have (transactionID, overseasEntityID, appData)
+ *
+ * @returns Promise< Resource<OverseasEntityCreated> | Resource<HttpStatusCode> | ApiErrorResponse >
+ */
+export const unauthorisedResponseHandler = async ( fnName: string, req: Request, session: Session, ...otherParams: any[] ) => {
+  let client = createOAuthApiClient(session);
+
+  let response = await client.overseasEntity[fnName](...otherParams);
```

Comment:
As the function is going to be for the overseas-entities-api only (using client.overseasEntity), i was wondering if this function should live in the overseas.entities.service.ts and rename the function to something like 'makeOverseasEntitiesApiCallWithRetry'

Reply:
Yes, agree. Even if it doesn't move, renaming the function name makes sense as that's what it is doing.

Reply:
I could make it more generic by adding the sdk object after the client call, and call it as you suggested makeApiCallWithRetryOnUnauthorisedResponse. Something like this and call the function

```
export const makeApiCallWithRetryOnUnauthorisedResponse = async ( sdkObjName: string, fnName: string, req: Request, session: Session, ...otherParams: any[] ) => {
  let client = createOAuthApiClient(session);

  let response = await client[sdkObjName][fnName](...otherParams);

  if (response && response.httpStatusCode === 401){

    const responseMsg = `Retrying ${fnName} call after unauthorised response`;
    logger.debugRequest(req, `${responseMsg} - ${JSON.stringify(response)}`);

    const accessToken = await refreshToken(req, session);
    logger.infoRequest(req, `New access token: ${accessToken}`);

    client = createOAuthApiClient(session);
    response = await client[sdkObjName][fnName](...otherParams);

  }

  return response;

};
```

Reply:
👍 not sure it needs to be more generic though as in this context (within OE web) it will always be 'overseasEntity', no?

Reply:
Ah right. So don't we need it on the POST call to create the OE too actually (for when save and resume feature is turned off)? Even when it's not turned off - when we call POST to create the initial OE with just the presenter info?

Basically, for all our remote calls from web to the OE API?

Reply:
The changes are already covering the POST and PUT for the OE calls

Reply:
Shouldn't it be done as part of this PR though? Not the removal of the feature flag but the redundant refresh call.

Reply:
Nope

Reply:
But this PR means that this line is now immediately redundant - hence why I think we remove it now.

Reply:
well, as I said it is part of the feature flag
```
    if (isActiveFeature(config.FEATURE_FLAG_ENABLE_REFRESH_TOKEN_29092022)) {
      const accessToken = await refreshToken(req, session);
      logger.infoRequest(req, `New access token: ${accessToken}`);
    }
```

Reply:
You do not like it, you disable the feature flag